expirationwatch/expiration_watcher.go

```diff
@@ -0,0 +1,162 @@
+package expirationwatch
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ocdogan/rbt"
+	log "github.com/sirupsen/logrus"
+)
+
+// ExpiredItem represents an expired item returned from the Watcher
+type ExpiredItem struct {
+	ExpirationTimeSeconds int64
+	Hash                  common.Hash
+}
+
+// Watcher watches the expiration of items
+type Watcher struct {
+	expiredOrders    chan []ExpiredItem
+	rbTreeMu         sync.RWMutex
+	rbTree           *rbt.RbTree
+	expirationBuffer time.Duration
+	ticker           *time.Ticker
+	isWatching       bool
+	wasStartedOnce   bool
+	mu               sync.Mutex
+}
+
+// New instantiates a new expiration watcher. An expiration buffer (positive or negative) can
+// be specified, causing items to be deemed expired some time before or after their expiration reaches current
+// UTC time. A positive expirationBuffer will make the item expire sooner then UTC, and a negative buffer after.
+func New(expirationBuffer time.Duration) *Watcher {
+	rbTree := rbt.NewRbTree()
+	return &Watcher{
+		expiredOrders:    make(chan []ExpiredItem, 10),
+		rbTree:           rbTree,
+		expirationBuffer: expirationBuffer,
+	}
+}
+
+// Add adds a new item identified by a hash to the expiration watcher
+func (w *Watcher) Add(expirationTimeSeconds int64, hash common.Hash) {
```

Comment:
Accepting the time as `int64` is unidiomatic and can lead to bugs/confusion because it is not always clear what unit of time is represented. It also restricts us unnecessarily to second precision (made sense when this package was used only for watching orders, but not when it is more generic). And finally, it doesn't allow us to take advantage of Go's builtin [monotonic clocks](https://golang.org/pkg/time/#hdr-Monotonic_Clocks) (without this, you can see bugs that occur whenever the clock is rolled back or jumps forward).

Can we change this type to `time.Duration`?

Reply:
These are actually UNIX timestamps passed around as `int64` because they are stored as Big.Ints in the `Order` struct type. We could change this to try `time.Time`, but `time.Duration` doesn't make sense here.

Reply:
Yes, sorry, I meant to suggest `time.Time`, not `time.Duration`. The `time` package makes a lot of considerations under the hood (e.g. timezones, monotonic clocks, different units of time) that we often don't get right when trying to do things manually. It's a huge win for preventing bugs and I think it would be a mistake not to use it.

> Perhaps we should refactor the Order type to represent ExpirationTimeSeconds using time.Time?

I don't think we need to do that. It's more convenient to make the `Order` type closely match the JSON-encoding and ABI-encoding for orders in the 0x protocol. However, the expiration watcher is not just for watching orders anymore so I think it should use the more generic `time.Time` type.

That said, you know better than I do how this generic expiration watcher is going to be used in the near future. If you think it's easier/faster to leave it as is, then I'm okay with it.

Reply:
Np, refactored to time.Time.